### ISSUE TITLE
 Intent to implement: amp-3d-player

### ISSUE THREAD

**Issue by kirillsalikhov** [ampproject/amphtml#14299]
We want to implement component for 3d model display.

Proposed behavior:
* it will allow to load and display 3d model in glTF format 
* it will allow rotate, zoom object

Example:
```
<amp-3d-player 
  width=”400”
  height=”500”
  src=”model-example.glb”>
</amp-3d-player>
```
Attributes:
* src - glb file path, required attribute
* alpha - true|false, canvas renders with transparent background, default true
* antialiasing - true|false, default true
* maxPixelRatio - force upper limit for pixel ratio, default to device pixel ratio 
* enableZoom - true|false, default true
* autoRotate - true|false, smoothly rotates 3d model around Y-axis, default false

For first version we are going to use Three.js library and stick to glb form of glTF.

**Comment by aghassemi** [comment]
@kirillsalikhov 

Sorry for the delay in getting back to you.

Few comments:

1- we normally use the `player` keyword for videos. Also `amp-3d-player` seems too generic, my understanding is that this extension would only support playing `glb` model files. My suggestion is `amp-3d-glb` or similar. We normally make these decisions in Design Review [meetings](https://github.com/ampproject/amphtml/blob/ebbd55e16c0908ec139243cd11d8c99e8da1bb3b/contributing/design-reviews.md). They are open to the public and you are welcomed and encouraged to present your proposal there. If you rather not, I can present on your behalf.

2- Regarding implementation, we have specific infra structure for 3P iframe. I recommend looking at `amp-mathml` code as reference for your implementation. Specifically https://github.com/ampproject/amphtml/blob/master/extensions/amp-mathml/0.1/amp-mathml.js and https://github.com/ampproject/amphtml/blob/master/3p/mathml.js

**Comment by kirillsalikhov** [comment]
Thanks for the replay.

1. We will `amp-3d-glTF`, if it's ok. 
I think we are not ready to present our proposal yet.

2. The big difference between 3d-player and other amp components involving third party js is that code that lives in iframe in other components is very consice. Which is not our case, we have 6 files and some of them quite large. Is it still ok to put them in `3p/`?

3. Another problem with the type-checking errors. We have about 400 of them) Most of these errors are from the fact that three.js is pulled from cdn, and not involved in build process at all. We tried to `yarn add` it and then use it in externs. It reduced the number from 400 to 80. We are okay with writing externs for the rest errors manually, (or even for all of them if it comes to that) but maybe there is some easier way to fix it?

**Comment by aghassemi** [comment]
1- sounds good.
2- Your code would need to run in an iframe as well. We can not allow three.js, etc.. run as first-party in AMP pages. If you look at components such as `amp-mathml`, `amp-ima-video`, `amp-twitter`, etc.. They have "two" files: 1- in 3P folder that runs "inside" iframe. This file can load external JS files (e.g. three.js) from other CDNs and then you can write whatever needed to make the model render, etc..  2- a `amp-x-component.js` file which essentially loads the iframe and passes attributes to it. It can also use post messages to communicate between the component and iframe.
3- To fix type checks, you would need to add the API surface of the third-party SDK's that you are using to `ad.extern.js` file (badly named, ad == 3p in this case). For instance `amp-mathml` used MathJax library loaded from a CDN and it adds it's API surface to that extern file like this:
```
//mathml.js
data.formula
var mathjax
mathjax.Hub
mathjax.Hub.Queue
window.MathJax
```

Hope this helps.

**Comment by kirillsalikhov** [comment]
1- We renamed component to amp-3d-glTF

2- We moved 6 files that are run in iframe to [amphtml/3p/3d-gltf/](https://github.com/webgears-3d/amphtml/tree/13c2de28ddc97082eebfed04a7567746e7852480/3p/3d-gltf) folder and reworked amp-3d-gltf component to use infra structure for 3P iframe.
Is it correct way?

**Comment by aghassemi** [comment]
@kirillsalikhov yes that approach is great. If someone those files are generic enough and available on CDNs (same way 3d is) you can also use the CDN version. Are these files written by yourself for this amp-component? If they are 3P, we need to ensure licenses match and are attributed properly.

I think a good next step is getting the code to a good state and submitting a PR and we can collaborate further on the PR.

**Comment by kirillsalikhov** [comment]
We found CDN for 2 of these files in 3p. Added license info for rest of them.

We are almost ready for PR, need some time for small refactor

**Comment by aghassemi** [comment]
Closed by #15147
